### ISSUE TITLE
A comparison of machine learning techniques for phishing detection

### ISSUE THREAD

**Issue by AndrewBreyen** [team3-cs373/resources#62]
### Title
A comparison of machine learning techniques for phishing detection

### URL
https://doi.org/10.1145/1299015.1299021

### Summary 
This study attempts to compare the predictive accuracy of six different classifiers, including Logistic Regression (LR), Classiﬁcationand Regression Trees (CART), Bayesian Additive Regression Trees (BART), Support Vector Machines (SVM), Random Forests (RF), and Neural Networks (NNet).

The data set was created with 1171 raw phishing emails, and 1718 legitimate emails. There were 43 variables used for training and testing the six classifiers. 

The methods that were used for classification methods include:
**Logistic Regression:** This performs well when relationship is approximately linear, poorly if complex non-linear relationships exist. Requires more statistical assumptions before being applied than other techniques. Affected by missing data.

**Classification and Regression Trees (CART):** Describes the conditional distribution of y given x. two components: a tree T with b terminal nodes, and a parameter vector Θ = (θ1, θ2, . . . , θb) where θi is associated with the ith terminal node. A binary tree is used to partition the predictor space recursively into distinct homogenous regions, where the terminal nodes of the tree correspond to the distinct regions. CART is ﬂexible in practice in the sense that it can easily model nonlinear or nonsmooth relationships. It has the ability of interpreting interactions among predictors. It also has great interpretability due to its binary structure. However,CART has several drawbacks such as it tends to overﬁt the data. In addition, since one big tree is grown, it is hard to account for additive eﬀects.

**Random Forests:** Many tree predictors are combined - each tree depends on the values of a random vector sampled independently. All trees in the forest have the same distribution. To construct a tree: N=number of training observations, P=number of variables (features) in the training set. In order to determine the decision node at a tree we choose k << p as the number of variables to be selected. We select a bootstrap sample from the n observations in the training set and use the rest of the observations to estimate the error of the tree in the testing phase. K variables are randomly chosen as a decision at a certain node and the best split is calculated based on that K. Random Forests can handle large numbers of variables, and during the forest building process, an internal unbiased estimate of error is developed. Random forests can estimate missing data well,, but there is a lack of reproducibility as the process of building the forest is random. 

**Neural Networks:** Structured as a set of interconnected neurons, which are used to send signals from one neuron to others. Interconnected neurons have different weights to enhance the delivery among neurons. Individual neurons are not powerful by themselves, but with others, they can become more powerful. The power of NNs come from the nonlinearity of the hidden neurons. In order to learn complex mappings, nonlinearity is required. NNs require some experience due to multiple local minima and delicate regularization is required. 

**Support Vector Machines (SVMs):** One of the most popular classifiers. SVMs find the optimal separating hyperplane between two classes by maximizing the margin between the classes closest points. Several drawbacks: require high computational power to train, sensitive to noisy data. 

**Bayesian Additive Regression Trees (BART):** Used to discover unknown relationship f between a continuous output Y and a P dimensional vector of inputs x = (x1, ..., xp). By using a sum-of-trees model, BART is vastly more ﬂexible than single tree models that can hardly account for additive eﬀects. Each tree component is actually a “weak learner”, which explains a small and diﬀerent part of the unknown relationship between the input and output. In addition, BART can easily incorporate high-order interaction eﬀects among three or more input variables, which can be diﬃcult to be captured by other additive models. Furthermore, BART does not require variables selection, which is preformed automatically as the trees are built.

Data Set: 2889 emails, 59.5% legitimate. 
Results: 
Error Rate when λ = 1:
1. (Most effective) NNet
2. SVM
3. BART
4. LR
5. CART
6. RF

 Error Rate when λ = 9:
1. RF
2. SVM
3. CART
4. NNet
5. BARt
6. LR

False Positive (FP) and False Negative (FN) Rate 
| |FP|FN|
| :------------- | :----------: | -----------: |
|LR| 04.89%|17.04%|
|CART| 07.68%| 12.93%|
|SVM| 07.92%| 17.26%|
|NNet| 05.85%| 21.72%|
|BART| 05.82%| 18.92%|
|RF| 08.29%| 11.12%|

![image](https://user-images.githubusercontent.com/7895581/106372100-f0d57380-6331-11eb-81d8-206280c8942d.png)

The results showed that, when legitimate and phishing emails are weighted equally, RF outperforms all other classiﬁers with an error rate of 07.72%, followed by CART, LR,BART, SVM, and NNet respectively. NNet achieved the worst error rate of 10.73%. Although RF outperformed all classiﬁers, it achieved the worst false positive rate of 08.29%. LR had the minimum false positive rate of 4.89%. When applying cost-sensitive measures, i.e. penalizing false positives 9 times more than false negatives, LR outperformed all classiﬁers achieving the minimum weighted error rate of 03.82% followed by BART, NNet, CART, SVM, and RF respectively. However, RF had the worst weighted error rate of 05.78% when λ = 9. 
### Citation
@inproceedings{10.1145/1299015.1299021,
author = {Abu-Nimeh, Saeed and Nappa, Dario and Wang, Xinlei and Nair, Suku},
title = {A Comparison of Machine Learning Techniques for Phishing Detection},
year = {2007},
isbn = {9781595939395},
publisher = {Association for Computing Machinery},
address = {New York, NY, USA},
url = {https://doi.org/10.1145/1299015.1299021},
doi = {10.1145/1299015.1299021},
abstract = {There are many applications available for phishing detection. However, unlike predicting spam, there are only few studies that compare machine learning techniques in predicting phishing. The present study compares the predictive accuracy of several machine learning methods including Logistic Regression (LR), Classification and Regression Trees (CART), Bayesian Additive Regression Trees (BART), Support Vector Machines (SVM), Random Forests (RF), and Neural Networks (NNet) for predicting phishing emails. A data set of 2889 phishing and legitimate emails is used in the comparative study. In addition, 43 features are used to train and test the classifiers.},
booktitle = {Proceedings of the Anti-Phishing Working Groups 2nd Annual ECrime Researchers Summit},
pages = {60–69},
numpages = {10},
keywords = {SVM, NNet, machine learning, random forests, CART, logistic regression, classification, BART, phishing},
location = {Pittsburgh, Pennsylvania, USA},
series = {eCrime '07}
}
